### ISSUE TITLE
Tokenizer for 2.2 is still extremely slow compared to 1.9

### ISSUE THREAD

**Issue by ibarshai** [explosion/spaCy#4370]
I used Spacy 1.9 in the past for tokenizing large quantities of small strings. When 2.0 came out, the tokenization speed for these smaller strings dropped dramatically. I believe tokenizer speed was targeted in 2.1 and was hoping to see more improvements on this front in 2.2. Is there an intent to address this in the future? Am I doing something wrong with disabling pipeline elements that's keeping the processing of these strings very heavy?

To illustrate my point, here is a comparison of tokenizing a small sentence with most of the pipeline disabled showing a 40x difference is speed:

### Spacy 1.9
```
import spacy
nlp = spacy.load('en', parser=False, tagger=False, entity=False)
s = 'this is a sentence to tokenize'
%timeit nlp(s)

83.3 µs ± 418 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

### Spacy 2.2
```
import spacy
nlp = spacy.load('en_core_web_sm', 
                 disable=['parser' 'tagger', 'entity', 'ner', 
                          'entity_linker', 'entity_ruler', 
                          'textcat', 'sentencizer', 
                          'merge_noun_chunks', 'merge_entities', 'merge_subtokens'])
s = 'this is a sentence to tokenize'
%timeit nlp(s)

3.27 ms ± 47.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

## My Environment
* **spaCy version:** 1.9.0
* **Platform:** Linux-4.13.0-45-generic-x86_64-with-debian-stretch-sid
* **Python version:** 3.6.2
* **Installed models:** en
--------
* **spaCy version:** 2.2.1
* **Platform:** Windows-10-10.0.18362-SP0
* **Python version:** 3.7.4
* **Installed models:** en_core_web_sm

**Comment by honnibal** [comment]
You're likely running into some cache effects. The en_core_web_sm model doesn't provide cached lexical entries, unlike the v1 models. Try the en_core_web_lg. The difference will be small regardless once you're tokenising more text.

You might also find it helps to disable the lexical attributes: nlp.vocab.lex_attr_getters = {}. Finally, be sure to benchmark with a more reasonable sample. You want to be timing functions that take at least 30 to 60 seconds. Otherwise you really can't conclude anything.

**Comment by ibarshai** [comment]
Thanks for the quick response! I tried the things you suggested and I'm actually still seeing very similar performance comparisons.

I took the text of [Ulysses](http://www.gutenberg.org/files/4300/4300-0.txt) and tried tokenizing it both as a whole text and after splitting by new lines. The results are:

```
with open('ulysses.txt', 'r') as f: text = f.read()
len(text) # 1513827 characters

text_split = [t for t in text.split('\n') if len(t)>0]
len(text_split) # 24767 lines
```
### Spacy 1.9:
```
import spacy
nlp = spacy.load('en', parser=False, tagger=False, entity=False)

%timeit nlp(text)
# 1.13 s ± 2.51 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit for t in text_split: nlp(t)
# 14.3 s ± 163 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
### Spacy 2.2:
I tried your suggestions both separately and together. It didn't seem to have any noticeable effect.

```
import spacy

nlp_sm = spacy.load('en_core_web_sm', 
                    disable=['parser' 'tagger', 'entity', 'ner', 
                             'entity_linker', 'entity_ruler', 'textcat', 
                             'sentencizer', 'merge_noun_chunks', 
                             'merge_entities', 'merge_subtokens'])

nlp_lg = spacy.load('en_core_web_lg', 
                    disable=['parser' 'tagger', 'entity', 'ner', 
                             'entity_linker', 'entity_ruler', 'textcat', 
                             'sentencizer', 'merge_noun_chunks', 
                             'merge_entities', 'merge_subtokens'])

nlp_sm.max_length = 2000000
nlp_lg.max_length = 2000000

#===== Timing With Default Lexical Attributes =====
# tokenizing of full text
%timeit nlp_sm(text)
%timeit nlp_lg(text)

# tokenizing of each line of text
%timeit for t in text_split: nlp_sm(t)
%timeit for t in text_split: nlp_lg(t)

#===== Timing With Disabled Lexical Attributes =====
nlp_sm.vocab.lex_attr_getters = {}
nlp_lg.vocab.lex_attr_getters = {}

# tokenizing of full text
%timeit nlp_sm(text)
%timeit nlp_lgtext)

# tokenizing of each line of text
%timeit for t in text_split: nlp_sm(t)
%timeit for t in text_split: nlp_lg(t)
```
To summarize the timings for 2.2:
```
                                 | sm model | lg model |
|:----------:|:-----------------:|:--------:|:--------:|
| whole text |  lex_attr default |  25.4 s  |  24.9 s  |
|            | lex_attr disabled |  24.5 s  |  24.6 s  |
| split text |  lex_attr default | 1min 38s | 1min 40s |
|            | lex_attr disabled |          |          |
```

So I'm still seeing quite a gap in performance. I realize that the new models do quite a bit more, but is there anything else I could do to get even slightly comparable performance?
